java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java

```diff
@@ -365,45 +360,56 @@ private <T extends IncrementalTask> boolean lockShard(Transaction tx,
   @Override
   public void runTask(final ShardedJobRunId jobId, final IncrementalTaskId taskId, final int sequenceNumber) {
     //acquire lock (allows this process to START potentially long-running work of task itself)
-    Transaction lockAcquisition = getDatastore().newTransaction();
 
-    try {
-      final ShardedJobStateImpl<? extends IncrementalTask> jobState = lookupJobState(lockAcquisition, jobId);
+    RetryExecutor.<Void>call(FOREVER_RETRYER, () -> {
+      Transaction lockAcquisition = getDatastore().newTransaction();
+      try {
+        final ShardedJobStateImpl<? extends IncrementalTask> jobState = lookupJobState(lockAcquisition, jobId);
 
-      if (jobState == null) {
-        log.info(taskId + ": Job is gone, ignoring runTask call.");
-        return;
-      }
+        if (jobState == null) {
+          log.info(taskId + ": Job is gone, ignoring runTask call.");
+          return null;
+        }
 
-      //taskState represents attempt of executing a slice of a shard of a sharded job
-      IncrementalTaskState taskState =
-        getAndValidateTaskState(lockAcquisition, taskId, sequenceNumber, jobState);
-      if (taskState == null) {
-        // some sort of error code happened
+        //taskState represents attempt of executing a slice of a shard of a sharded job
+        IncrementalTaskState taskState =
+          getAndValidateTaskState(lockAcquisition, taskId, sequenceNumber, jobState);
+        if (taskState == null) {
+          // some sort of error code happened
+
+          // seems like getAndValidationTaskState has potential side-effects, which need to be committed
+          lockAcquisition.commit();
+          return null;
+        }
 
-        // seems like getAndValidationTaskState has potential side-effects, which need to be committed
-        lockAcquisition.commit();
-        return;
+        if (lockShard(lockAcquisition, taskState)) {
+          // committing here, which forces acquisition of lock ...
+          lockAcquisition.commit();
+
+          // actual task execution
+          runAndUpdateTask(jobState.getShardedJobId(), taskId, sequenceNumber, jobState, taskState);
```

Comment:
I think this was triggering the DatastoreException, not being caught, so reaching the servlet execution